Rolldown-Vite

sapphi-red opened the pull request:
**Rolldown-Vite will be merged to Vite since Vite 8 beta**

This PR will track rolldown-vite 7.x.

<details>

<summary>Old information</summary>

### Description

#### Current Status

- Vite repo test status
  - ✅ unit tests
  - e2e tests
    - ✅ external: changed the test for intended behavior change
    - ⚠️ legacy: some manifest related tests are skipped for now
    - ⚠️ object-hooks: parallel hooks does not run in parallel
    - ✅ resolve: skipped some browser field related tests as those cases were only supported by esbuild and Vite, also skipped some test that does fallback based on the file content which is Vite only and I believe it should not be done
- ecosystem-ci (see https://github.com/vitejs/vite-ecosystem-ci/blob/rolldown-vite/README-temp.md) 

#### additional features
- tsconfig paths support (https://github.com/vitejs/rolldown-vite/pull/277)
  - When `resolve.tsconfigPaths: true` is set, resolution respecting tsconfig paths is supported
  - This option requires `experimental.enableNativePlugin` to be set to `'resolver'` or `true`
  - Currently this feature only supports basic cases. If the `tsconfig.json` uses `references`/`include`/`exclude`, it won't work as expected.
- automatic `emitDecoratorMetadata: true` support (https://github.com/vitejs/rolldown-vite/pull/315)
  - Since Oxc supports `emitDecoratorMetadata: true`, rolldown-vite will enable that transform if it is enabled in tsconfig.json
- [`moduleType`](https://rolldown.rs/guide/in-depth/module-types) is supported in hook filters (experimental)

#### Intended behavior differences
- different `import default from './foo.cjs'` behavior (see https://rolldown.rs/in-depth/bundling-cjs#ambiguous-default-import-from-cjs-modules)
- `esbuild` needs to be installed manually in the following cases:
  - when using `build.minify: 'esbuild'`
  - when using a plugin using `transformWithEsbuild` function
- passing the same browser with multiple versions of it to `build.target` errors: esbuild selects the latest version of it, but that's probably not what you intend
- https://github.com/oxc-project/oxc/issues/7641
- require is never transformed into import, users needs to inject a polyfill on their own, see https://rolldown.rs/in-depth/bundling-cjs#require-external-modules
- [`build.rollupOptions.output.manualChunks`](https://rollupjs.org/configuration-options/#output-manualchunks) is deprecated. please use [`build.rollupOptions.output.advancedChunks`](https://rolldown.rs/guide/features#advanced-chunks) instead when possible.
- `structuredClone(bundle)` errors with `DataCloneError: #<Object> could not be cloned.`. please clone it with `structuredClone({ ...bundle })` (https://github.com/vitejs/rolldown-vite/issues/128)
- assigning to `bundle[foo]` in `generateBundle` hook is not supported. It is discouraged by Rollup as well.
- the `bundle` object reference is not shared across the hooks (https://github.com/vitejs/rolldown-vite/issues/410)
- (native plugin): if the package has both `browser` field and `module` field, the non-native resolver checked both fields and picked the one that looks like an ESM file. this behavior is not implemented in the native resolver as it was confusing and only legacy packages have this problem.

#### Known missing features

- [`output.format: 'system'`](https://rollupjs.org/configuration-options/#output-format) (https://github.com/rolldown/rolldown/issues/2387)
- [`output.format: 'amd'`](https://rollupjs.org/configuration-options/#output-format) (https://github.com/rolldown/rolldown/issues/2528)
- native decorator support (https://github.com/oxc-project/oxc/issues/9170)
- incomplete Typescript legacy namespace support (https://github.com/oxc-project/oxc/issues/14227)
- `shouldTransformCachedModule` is not supported yet (https://github.com/rolldown/rolldown/issues/4389)
- `resolveImportMeta` hook (https://github.com/rolldown/rolldown/issues/1010)
- see https://github.com/rolldown/rolldown/issues/4973 for native plugin compat status

</details>

christianjuth commented:
Found this PR in the Vite docs. Curious how stable `rolldown-vite` is. On a scale of "Next.js thinks this is stable" to "this will actually work 99% of the time", were are we at?

sapphi-red commented:
> Found this PR in the Vite docs. Curious how stable `rolldown-vite` is. On a scale of "Next.js would call this is stable" to "this will actually work 99% of the time", were are we at?

I'd say it would work for basic cases as it is passing most tests through out the ecosystem.

jclaveau commented:
> Found this PR in the Vite docs. Curious how stable `rolldown-vite` is. On a scale of "Next.js would call this is stable" to "this will actually work 99% of the time", were are we at?

You can test it within minutes in your projects following this https://main.vite.dev/guide/rolldown#how-to-try-rolldown

In my case (using roughly 10 Vite plugins) the main blocking point was `vite-plugin-legacy` which doesn't work due to the lack of System JS output
I also had an odd Rust error while running the build more than one time (So it seems to be a cache issue).

But damn, the build is blazing fast!

sapphi-red commented:
> > define: per-environment config is not supported yet
> 
> Is there an open issue for this?

No, there isn't. But checking the code, actually it should work. Feel free to create one if it's not working.

sapphi-red committed:
fix: make `viteMetadata` and `modules` enumerable (#261)

sapphi-red committed:
fix: set `viteMetadata` for return value of `build` function (#262)

sapphi-red committed:
release: v6.3.21

sapphi-red committed:
chore: merge commit '86b5e0030bf204f8f2db0cf8ee895ab3ecf154b8' into rolldown-vite

sapphi-red committed:
chore: merge commit 'a37ac836ac4da8e854d98c65450f12acb921aa98' into rolldown-vite

sapphi-red committed:
Merge commit '61b6b96b191c6071b9c574ad4c795f97f2646f18' into rolldown-vite

sapphi-red committed:
chore: merge commit 'a33d0c7d65d9fff9acd5de0cf3c4d371297b3990' into rolldown-vite

sapphi-red committed:
chore: merge commit '24692dbeccd148a60f563bc96e12e3609e5ba642' into rolldown-vite

sapphi-red commented:
I'll lock this PR as comments on this PR will be difficult to track. Feel free to create an issue or discussion if you need to comment.

sapphi-red committed:
fix: merge conflict mistake

sapphi-red committed:
fix: merge conflict mistake

sapphi-red committed:
chore: merge commit '9c6cbe602187644887b0a4342de6b486c8f60f4d' into rolldown-vite

sapphi-red committed:
chore: merge main

sapphi-red committed:
fix(native-plugin): run optimizer correctly (#265)

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
fix: importing an optional peer dep should throw an runtime error (#167)

sapphi-red committed:
perf: set filter for import analysis build plugin (#266)

sapphi-red committed:
fix: set viteMetadata for `writeBundle` hook (#263)

sapphi-red committed:
release: v7.0.0-alpha.0

sapphi-red committed:
fix: resolve oxc runtime from Vite directory correctly (#269)

sapphi-red committed:
fix(css): set moduleType in css-post plugin (#272)

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
fix(json): require(json) should return the content (#268)

sapphi-red committed:
release: v7.0.0-beta.0

sapphi-red committed:
chore: merge main

shulaoda committed:
feat(native-plugin): fallback to js alias plugin when alias entries include `customResolver` (#280)

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.0.0

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix docs build

shulaoda committed:
feat: update rolldown (#279)

sapphi-red committed:
release: v7.0.1

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
release: v7.0.2

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
chore: merge main

sapphi-red committed:
feat: add `resolve.tsconfigPaths` option to resolve with tsconfig paths (#277)

sapphi-red committed:
release: v7.0.3

shulaoda committed:
feat(native-plugin): avoid presetting `process.env.NODE_ENV` when platform is `browser` (#290)

Co-authored-by: 翠 <green@sapphi.red>

shulaoda committed:
feat: update rolldown and support `resolveSubpathImports` for `oxcResolvePlugin` (#287)

shulaoda committed:
feat(native-plugin): use js define plugin in dev environment (#298)

shulaoda committed:
feat(native-plugin): pass `decodedBase` to `nativeWasmHelperPlugin` (#300)

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.0.4

sapphi-red committed:
fix: correct "Unexpected IIFE format" shown for UMD output

sapphi-red committed:
fix: accept umd with only default export (#305)

sapphi-red committed:
test: enable oxc runtime injection in lib playground

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: preview release plugin-legacy

sapphi-red committed:
chore: preview release plugin-legacy

sapphi-red committed:
feat: make oxc runtime helpers name mangle-able (#306)

sapphi-red committed:
feat: plugin legacy (#293)

Co-authored-by: Hiroshi Ogawa <hi.ogawa.zz@gmail.com>

sapphi-red committed:
chore: remove `ROLLDOWN_OPTIONS_VALIDATION=loose` from vitepress config

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: reduce diff

shulaoda committed:
feat: update rolldown and support custom configs for native dynamic import vars plugin (#310)

sapphi-red committed:
fix: inject oxc runtime helpers correctly when multiple helpers are required (#311)

sapphi-red committed:
release: v7.0.5

sapphi-red committed:
fix(legacy): fix remaining issues (#312)

sapphi-red committed:
feat: enable `treeshake.commonjs` by default (#313)

sapphi-red committed:
feat: enable decorator-metadata transform automatically (#315)

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.0.6

hi-ogawa committed:
fix: disable refresh for non-jsx without react import (#318)

sapphi-red committed:
perf(legacy): set some babel options

sapphi-red committed:
chore: merge main

sapphi-red committed:
refactor: use Rolldown's `@oxc-project/runtime` (#321)

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
feat: enable `output.minifyInternalExports` by default (#322)

sapphi-red committed:
release: v7.0.7

sapphi-red committed:
chore: merge main

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
release: v7.0.8

sapphi-red committed:
chore: update some "rollup" -> "rolldown"

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
release: v7.0.9

sapphi-red committed:
feat: warn when `esbuild: false` is set but `oxc: false` is not set

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: merge main

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
release: v7.0.10

sapphi-red committed:
fix(worker): match worker import meta more quickly (#336)

shulaoda committed:
feat(native-plugin): inject css and assets for native manifest plugin (#335)

sapphi-red committed:
fix(lib): keep annotation comments for es output (#337)

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.0.11

sapphi-red committed:
fix: reset viteMetadata in watch mode (#342)

sapphi-red committed:
test: enable optimize-deps cjs test case

sapphi-red committed:
chore: update rolldown-plugin-dts

for https://github.com/sxzz/rolldown-plugin-dts/issues/66

sapphi-red committed:
feat: bump rolldown (#344)

sapphi-red committed:
feat(native-plugin): output warn / debug logs from resolver (#343)

sapphi-red committed:
release: v7.0.12

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix type error

sapphi-red committed:
fix: apply https://github.com/vitejs/vite/pull/20503

sapphi-red committed:
feat: bump rolldown

sapphi-red committed:
release: v7.1.0-beta.0

sapphi-red committed:
feat: enable native resolver by default (#346)

sapphi-red committed:
refactor: introduce `*.rolldownOptions` and deprecate `*.rollupOptions` (#348)

sapphi-red committed:
fix: watch the config file itself (#354)

sapphi-red committed:
release: v7.1.0-beta.1

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.1.0-beta.2

sapphi-red committed:
chore: update JSDoc of `experimental.enableNativePlugin` option

refs https://github.com/vitejs/rolldown-vite/pull/346

sapphi-red committed:
refactor: deprecate esbuild / optimizeDeps.esbuild options (#355)

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: merge main

sapphi-red committed:
fix(legacy): use remove conditions for normal Vite

sapphi-red committed:
release: v7.1.0

hi-ogawa committed:
fix: respect `rollupOptions.platform` options (#361)

shulaoda committed:
feat: update rolldown (#366)

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix typecheck

sapphi-red committed:
release: v7.1.1

sapphi-red committed:
feat: enable all native plugins by default (#168)

Co-authored-by: shulaoda <165626830+shulaoda@users.noreply.github.com>

sapphi-red committed:
release: v7.1.2

sapphi-red committed:
feat: warn if a plugin specified both `rollupOptions` and `rolldownOptions` (#377)

sapphi-red committed:
feat: update rolldown (#379)

hi-ogawa committed:
fix: merge tsconfig jsx options even when `oxc.jsx.runtime` is set  (#376)

sapphi-red committed:
release: v7.1.3

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix js plugin test

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.1.4

sapphi-red committed:
chore: merge main

shulaoda committed:
fix(native-plugin): generate a more readable manifest (#388)

sapphi-red committed:
feat: update rolldown (#389)

sapphi-red committed:
release: v7.1.5

sapphi-red committed:
feat: enable `output.topLevelVar` by default (#380)

shulaoda committed:
fix(oxc): respect `oxc.typescript.onlyRemoveTypeImports` if it's set (#394)

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
fix: `build.minify: 'esbuild'` + native plugins were not working (#397)

sapphi-red committed:
release: v7.1.6

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.1.7

sapphi-red committed:
chore: support ESM rolldown

sapphi-red committed:
feat: enable `optimization.inlineConst: { mode: 'smart' }` by default (#399)

sapphi-red committed:
chore: update typos

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: update type build

sapphi-red committed:
feat: update rolldown (#400)

sapphi-red committed:
release: v7.1.8

sapphi-red committed:
fix: enable minify other than codegen for ES lib mode

fixes https://github.com/vitejs/rolldown-vite/issues/390

IWANABETHATGUY committed:
chore: bump `rolldown-plugin-dts` (#409)

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
release: v7.1.9

sapphi-red committed:
feat: update rolldown (#415)

sapphi-red committed:
release: v7.1.10

hi-ogawa committed:
fix(optimizer): avoid mutating `optimizeDeps.rollupOptions.transforms` for stable optimizer hash (#417)

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix typecheck

sapphi-red committed:
release: v7.1.11

sapphi-red committed:
chore: merge main

sapphi-red committed:
feat: update rolldown (#423)

sapphi-red committed:
release: v7.1.12

sapphi-red committed:
feat(hmr): skip self-imports for `import.meta.hot.invalidate`

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
release: v7.1.13

sapphi-red committed:
chore: merge main

sapphi-red committed:
fix: use readFileSync to avoid "too many open files"

sapphi-red committed:
perf: use single regex for `isEntirelyImport`

sapphi-red committed:
feat: update rolldown

sapphi-red committed:
feat: set `generatedCode: 'es2015'`

sapphi-red committed:
release: v7.1.14

sapphi-red committed:
fix(oxc): set typescript jsx options

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.1.15

sapphi-red committed:
chore: merge main

sapphi-red committed:
fix: set transform target (#441)

sapphi-red committed:
feat: update rolldown (#443)

sapphi-red committed:
release: v7.1.16

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix type error

sapphi-red committed:
feat: update rolldown

actually it was updated in 9cf9481bc285b52359d018d7568a0d74ff1a02a3
https://github.com/rolldown/rolldown/releases/tag/v1.0.0-beta.43

sapphi-red committed:
release: v7.1.17

sapphi-red committed:
refactor: revert #313

`treeshake.commonjs` is now enabled by default by rolldown

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.1.18

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix types

sapphi-red committed:
feat: update rolldown (#456)

sapphi-red committed:
fix: resolve default export consistently (#444)

sapphi-red committed:
refactor: revert #322
`output.minifyInternalExports` is now enabled by default by rolldown

sapphi-red committed:
refactor: avoid deprecated rolldown options (e.g. top level `define`)

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix type

sapphi-red committed:
release: v7.1.19

sapphi-red committed:
feat(legacy): add`build.rolldownOptions.output.sourcemapBaseUrl` support

sapphi-red committed:
test: enable temporary disabled tests

sapphi-red committed:
fix(optimizer): pass target to optimizer

sapphi-red committed:
feat: update rolldown (#467)

sapphi-red committed:
feat: add legacy inconsistent cjs interop (#464)

sapphi-red committed:
fix(deps): downgrade commonjs plugin to 28.0.6 to avoid rollup/plugins#1909 (#20988)

sapphi-red committed:
release: v7.1.20

sapphi-red committed:
fix(resolve): don't modify resolve options (#468)

sapphi-red committed:
chore: merge main

sapphi-red committed:
chore: fix typecheck

sapphi-red committed:
chore: fix merge error

sapphi-red committed:
release: v7.2.0-beta.0

sapphi-red committed:
fix: trigger watchChange hook for all environments for native resolver (#463)

sapphi-red committed:
fix: set inferred moduleType for virtual modules (#470)

sapphi-red committed:
fix(optimizer): support `namespace: 'file'` filter in esbuild plugin compat (#479)

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.2.0-beta.1

sapphi-red committed:
feat: update rolldown (#481)

sapphi-red committed:
release: v7.2.0-beta.2

sapphi-red committed:
fix(optimizer): sync `optimizeDeps.rollupOptions` and `optimizeDeps.rolldownOptions` (#482)

sapphi-red committed:
feat: update rolldown (#488)

sapphi-red committed:
feat: expose `minify` function (#489)

sapphi-red committed:
release: v7.2.0-beta.3

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.2.0

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.2.1

sapphi-red committed:
feat: convert `esbuild.jsxSideEffects` to `oxc.jsx.pure` (#493)

sapphi-red committed:
fix(optimizer): convert to `optimizeDeps.rolldownOptions.transform.define` from `optimizeDeps.esbuildOptions.define` (#494)

sapphi-red committed:
fix: set `legalComments: 'none'` to align with Vite (#497)

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.2.2

sapphi-red committed:
feat: update rolldown (#502)

sapphi-red committed:
release: v7.2.3

sapphi-red committed:
feat!: change `minify` to be async (#505)

sapphi-red committed:
feat: expose `parse` / `parseAsync` (#503)

sapphi-red committed:
fix: don't set `ssr.optimizeDeps.rollupOptions` compat if `ssr` option is not set (#506)

sapphi-red committed:
refactor: runtime deprecate `optimizeDeps.rollupOptions` (#507)

sapphi-red committed:
release: v7.2.4

sapphi-red committed:
feat: update rolldown to 1.0.0-beta.50 (#508)

sapphi-red committed:
release: v7.2.5

sapphi-red committed:
docs: migration guide (#469)

Co-authored-by: Alexander Lichter <github@lichter.io>

shulaoda committed:
feat: update rolldown to 1.0.0-beta.51 (#512)

sapphi-red committed:
release: v7.2.6

cpojer committed:
fix: esbuild plugin converter with empty JS files. (#516)

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.2.7

sapphi-red committed:
fix: `inconsistentCjsInterop: true` should affect dynamic imports as well (#521)

sapphi-red committed:
chore: merge main

shulaoda committed:
feat: update rolldown to 1.0.0-beta.52 (#522)

sapphi-red committed:
release: v7.2.8

sapphi-red committed:
chore: merge main

sapphi-red committed:
release: v7.2.9

sapphi-red committed:
test: add nested tsconfig paths test (#527)

sapphi-red committed:
feat: deprecate `build.commonjsOptions` (#530)

sapphi-red committed:
docs: update content for beta (#529)

sapphi-red committed:
docs: remove rolldown page (#528)

sapphi-red committed:
chore: merge main

sapphi-red committed:
test: remove failure expected cases (#536)

cpojer committed:
docs: update migration guide. (#538)

shulaoda committed:
feat: update rolldown to 1.0.0-beta.53 (#537)

Co-authored-by: sapphi-red <49056869+sapphi-red@users.noreply.github.com>

sapphi-red committed:
test: add test for https://github.com/vitejs/rolldown-vite/issues/523 (#524)

sapphi-red committed:
fix: disable resolver cache when watcher is disabled (#471)

sapphi-red committed:
release: v7.2.10
